dds/DCPS/transport/tcp/TcpReconnectTask.cpp

```diff
@@ -6,48 +6,93 @@
  */
 
 #include "Tcp_pch.h"
-#include "TcpReconnectTask.h"
-#include "TcpConnection.h"
-#include "TcpSendStrategy.h"
+#include "dds/DCPS/transport/tcp/TcpReconnectTask.h"
+#include "dds/DCPS/transport/tcp/TcpConnection.h"
+#include "dds/DCPS/transport/tcp/TcpTransport.h"
 #include "dds/DCPS/transport/framework/EntryExit.h"
 
+#include "ace/Reverse_Lock_T.h"
+
 OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
 
-OpenDDS::DCPS::TcpReconnectTask::TcpReconnectTask(
-  OpenDDS::DCPS::TcpConnection* connection)
-  : connection_(connection)
+namespace OpenDDS {
+namespace DCPS {
+
+TcpReconnectTask::TcpReconnectTask() : connection_(), mutex_(), cv_(mutex_), shutdown_(false)
 {
-  DBG_ENTRY_LVL("TcpReconnectTask","TcpReconnectTask",6);
+  DBG_ENTRY_LVL("TcpReconnectTask", "TcpReconnectTask", 6);
   if (open()) {
     ACE_ERROR((LM_ERROR,
                ACE_TEXT("(%P|%t) ERROR: Reconnect task failed to open : %p\n"),
                ACE_TEXT("open")));
   }
 }
 
-OpenDDS::DCPS::TcpReconnectTask::~TcpReconnectTask()
+TcpReconnectTask::~TcpReconnectTask()
+{
+  DBG_ENTRY_LVL("TcpReconnectTask", "~TcpReconnectTask", 6);
+  shutdown();
+  wait();
+}
+
+void TcpReconnectTask::reconnect(TcpConnection_rch connection)
+{
+  DBG_ENTRY_LVL("TcpReconnectTask", "reconnect", 6);
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+  if (!shutdown_ && !connection_) {
+    connection_ = connection;
+    activate(1);
+  }
+}
+
+bool TcpReconnectTask::active() {
+  ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mutex_, false);
+  return connection_;
+}
+
+void TcpReconnectTask::shutdown() {
+  DBG_ENTRY_LVL("TcpReconnectTask", "shutdown", 6);
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+  shutdown_ = true;
```

Comment:
Should this wake up a `wait_complete()`?

Reply:
The goal here was to have connection_ act as the single indicator for whether work was scheduled to be done, and so even if shutdown() has been called & shutdown_ is set, wait_complete would still want to wait until connection_ is no longer set.